### DISCUSSION TITLE
need help with getter function

### DISCUSSION THREAD

**Discussion by fredbouin** [alpinejs/alpine#3525] [1. Help] [answered]
in the code below I am calculating a price based off table_sizes and their associated values. I am also trying to add a button for an extra long option, my issue is that I don't want 'long' to have a fixed price, I need the extra amount to be dependent on what table_size is selected. So what I'm attempting is making 'long' a quarter of the selected value.

eg. 6p = 1000, 6p(long) = 1250, 8p = 1100 , 8p(long) = 1375

In size_mods I have `reg: 0`, and I put a getter function for long but I'm pretty sure I can't do that and that's where my problem is, but I'm not sure what a good approach would be. Maybe taking the getter out of size_mods... I'm not sure how to refer back to it after.

```
<div x-data="{	
    base: 1000,
    table_sizes: {
        '6p': 0,
        '8p': 100,
        '10p': 300,
    },
    table_size: '6p',
    size_mods: {
        reg: 0,
        get long() { return this.table_sizes[this.table_size] / 4; },
    },
    size_mod: 'reg',
         get price() {
             return (this.base + this.table_sizes[this.table_size] + this.size_mods[this.size_mod])
         },	
    }"
>
    <div>
        <button @click="table_size = '6p'" :class="{ 'underline' : table_size == '6p' }">
            6p
        </button>
        <button @click="table_size = '8p'" :class="{ 'underline' : table_size == '8p' }">
            8p 
        </button>
        <button @click="table_size = '10p'" :class="{ 'underline' : table_size == '10p' }">
            10p
        </button>
     </div>
    <div>
        <button @click="size_mod = 'reg'" :class="{ 'underline' : size_mod == 'reg' }">
            reg
        </button>
        <button @click="size_mod = 'long'" :class="{ 'underline' : size_mod == 'long' }">
            long
        </button>
    </div>
    <span x-text="price"></span>
</div>
```

**Comment by trych** [comment, marked as answer]
I think you could simplify this considerably by doing this:

```js
<div x-data="{  
    base: 1000,
    table_sizes: {
        '6p': 0,
        '8p': 100,
        '10p': 300,
    },
    long: false,
    table_size: '6p',
    price: 0
}"

    x-effect="price = long ? (base + table_sizes[table_size]) * 1.25 : base + table_sizes[table_size]"
>
    <div>
        <button @click="table_size = '6p'" :class="{ 'underline' : table_size == '6p' }">
            6p
        </button>
        <button @click="table_size = '8p'" :class="{ 'underline' : table_size == '8p' }">
            8p 
        </button>
        <button @click="table_size = '10p'" :class="{ 'underline' : table_size == '10p' }">
            10p
        </button>
     </div>
    <div>
        <button @click="long = false" :class="{ 'underline' : !long }">
            reg
        </button>
        <button @click="long = true" :class="{ 'underline' : long }">
            long
        </button>
    </div>
    <span x-text="price"></span>
</div>
```

I haven't tested this (next time a CodePen or something to test this would be helpful), but it should do what you want. There is a simple boolean for `long` and then most importantly there is an `x-effect` directive that re-evaluates and recalculates the price whenever one of the variables within `x-effect` changes.

**Reply by fredbouin** [reply]
thanks for taking the time, here is a codepen of my previous example:  https://codepen.io/fredbouin/pen/zYmvZgM
what you shared isn't currently working but I will spend some time with it, I had a feeling x-effect was an option here.

**Reply by trych** [reply]
You're right. I forgot to include the price in the `x-data` and forgot to close the curly braces. I changed the code above, it should work now.

**Reply by fredbouin** [reply]
great yes this works I think this approach should work well with the rest of the project. thanks!

**Comment by ekwoka** [comment]
I think the getter option is actually better than the x-effect, but definitely needs some improved organization.

It's easily to abstract the component to Alpine.data with the getter.

**Reply by fredbouin** [reply]
I would be interested in seeing both approaches if you wouldn't mind elaborating. What I was mainly wondering with the getter is how to do this: 
```
size_mods: {
        reg: 0,
        get long() { return this.table_sizes[this.table_size] / 4; },
    },
```

**Reply by ekwoka** [reply]
Is there an issue with how you did it there?

ah right...the nested object issue....

**Reply by SimoTod** [reply]
also, be careful that long is a reserved word in Javascript, I don't think you can use it in a getter in that way

**Reply by trych** [reply]
Now I am curious, too. How would a working implementation using the getter option look? How and why is it better than the `x-effect` option?

**Reply by SimoTod** [reply]
It's easier to just store 0.5 in size_mods and do all your calculation in getPrice, IMO. For example, a working implementation is
```
x-data="{	
  base: 1000,
  table_sizes: {
    '6p': 0,
    '8p': 100,
    '10p': 300,
  },
  table_size: '6p',
  size_mods: {
    reg: 0,
    long: 0.5
  },
  size_mod: 'reg',
  get price() {
    return (this.base + this.table_sizes[this.table_size] + (this.table_sizes[this.table_size]*this.size_mods[this.size_mod]))
  },	
}"
```
x-effect is moving an important logic of your component to a separate tag while having everything x-data allows you to easily abstract your logic into an Alpine.data definition in case you ever need it. It may not be a bigger concern for you though

**Reply by ekwoka** [reply]
What would make it better is just the encapsulation.

keeping similar structure you could do

```js
{	
    base: 1000,
    table_sizes: {
        '6p': 0,
        '8p': 100,
        '10p': 300,
    },
    table_size: '6p',
    size_mods: {
        reg: 1,
        long: 1.25,
    },
    size_mod: 'reg',
         get price() {
             return this.base + this.table_sizes[this.table_size] * this.size_mods[this.size_mod]
         },	
    }
  ```

**Reply by trych** [reply]
@ekwoka Thanks! But how/where do I set the price then? Wouldn't I still need an `x-effect` or something similar?

**Reply by ekwoka** [reply]
@trych It's a getter. Why would you need to set it?

**Reply by trych** [reply]
Sorry, you're right had a syntax error when I tried the setup. The getter is sufficient. Thanks again!

**Reply by fredbouin** [reply]
oh ... obviously I should have done it like this right from the get go, this makes 100% more sense than the direction I was going

**Comment by fredbouin** [comment]
For the sake of learning something though, I'm still curious why my first attempt didn't work. I've realized with a few tests that the getter for long() doesn't work when nested inside of size_mods and referencing table_size in the calculation. It works if it's a simple calculation like 100/4 though, or it also works if referencing table_size but not inside of size_mods.

here is a codepen
https://codepen.io/fredbouin/pen/QWZjYXV

any thoughts? could my first approach have worked at all?

**Reply by SimoTod** [reply]
No, it couldn't. Update your code to `get long() { console.log(this); return (this.table_size / 4) },`
It's just the way javascript works, the getter scope is always the current object.

**Reply by fredbouin** [reply]
ah got it thanks for that
